File: kanonak-wire/csharp/src/Kanonak.Wire/WireReader.cs
using System;
using System.Text;

namespace Kanonak.Wire
{
    /// <summary>
    /// A bounds-checked cursor over an immutable byte buffer. Never copies.
    ///
    /// <para>The C# half of the Kanonak wire kernel's read side
    /// (<c>kanonak.org/wire-form</c>, <c>wireFormatVersion "1"</c>).
    /// <see cref="Bytes(int)"/> and <see cref="Rest"/> return
    /// <see cref="ReadOnlyMemory{T}"/> views into the source buffer, never
    /// copies. Fail-loud contract: every failure is a <see cref="WireError"/>
    /// stating what was expected, what was found, and where.</para>
    /// </summary>
    public sealed class WireReader
    {
        private static readonly UTF8Encoding StrictUtf8 = new UTF8Encoding(
            encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);

        private const string HexLower = "0123456789abcdef";

        private readonly byte[] _buf;
        private int _pos;

        public WireReader(byte[] buf)
        {
            if (buf == null)
            {
                throw new ArgumentNullException(nameof(buf));
            }
            _buf = buf;
        }

        /// <summary>Count of unread bytes.</summary>
        public int Remaining => _buf.Length - _pos;

        private void Need(int n, string context)
        {
            if (_buf.Length - _pos < n)
            {
                throw WireError.Truncated(n, _buf.Length - _pos, _pos, context);
            }
        }

        public byte U8()
        {
            Need(1, "u8");
            return _buf[_pos++];
        }

        public ushort U16Be()
        {
            Need(2, "u16be");
            ushort v = (ushort)((_buf[_pos] << 8) | _buf[_pos + 1]);
            _pos += 2;
            return v;
        }

        public uint U32Be()
        {
            Need(4, "u32be");
            int p = _pos;
            uint v = ((uint)_buf[p] << 24) | ((uint)_buf[p + 1] << 16) | ((uint)_buf[p + 2] << 8) | _buf[p + 3];
            _pos += 4;
            return v;
        }

        /// <summary>Exactly n bytes as a zero-copy view of the source buffer.</summary>
        public ReadOnlyMemory<byte> Bytes(int n)
        {
            Need(n, $"bytes({n})");
            var v = new ReadOnlyMemory<byte>(_buf, _pos, n);
            _pos += n;
            return v;
        }

        /// <summary>16 bytes as a lowercase hyphenated UUID string. Any 16 bytes are legal.</summary>
        public string Uuid()
        {
            Need(16, "uuid");
            var chars = new char[36];
            int ci = 0;
            for (int i = 0; i < 16; i++)
            {
                if (i == 4 || i == 6 || i == 8 || i == 10)
                {
                    chars[ci++] = '-';
                }
                byte b = _buf[_pos + i];
                chars[ci++] = HexLower[b >> 4];
                chars[ci++] = HexLower[b & 0x0f];
            }
            _pos += 16;
            return new string(chars);
        }

        /// <summary>n bytes decoded as STRICT UTF-8. Bounds are checked before validity.</summary>
        public string Utf8(int n)
        {
            int start = _pos;
            Bytes(n);
            try
            {
                return StrictUtf8.GetString(_buf, start, n);
            }
            catch (DecoderFallbackException)
            {
                _pos = start; // the read did not take effect
                throw WireError.InvalidUtf8(start, $"utf8({n})");
            }
        }

        /// <summary>u16be length L, then exactly L bytes (zero-copy view).</summary>
        public ReadOnlyMemory<byte> LenPrefixedBytes16()
        {
            int start = _pos;
            Need(2, "lenPrefixedBytes16");
            int declared = (_buf[_pos] << 8) | _buf[_pos + 1];
            int remainingAfterLength = _buf.Length - _pos - 2;
            if (declared > remainingAfterLength)
            {
                throw WireError.LengthOverrun(declared, remainingAfterLength, start, "lenPrefixedBytes16");
            }
            _pos += 2;
            var v = new ReadOnlyMemory<byte>(_buf, _pos, declared);
            _pos += declared;
            return v;
        }

        /// <summary>All remaining bytes (possibly empty) as a zero-copy view. Never errors.</summary>
        public ReadOnlyMemory<byte> Rest()
        {
            var v = new ReadOnlyMemory<byte>(_buf, _pos, _buf.Length - _pos);
            _pos = _buf.Length;
            return v;
        }

        /// <summary>Errors <c>TrailingBytes</c> if any bytes remain.</summary>
        public void ExpectEnd()
        {
            int count = _buf.Length - _pos;
            if (count > 0)
            {
                throw WireError.TrailingBytes(count, _pos);
            }
        }
    }
}

File: kanonak-wire/java/src/main/java/org/kanonak/wire/WireReader.java
package org.kanonak.wire;

import java.nio.ByteBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CharsetDecoder;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;

/**
 * A bounds-checked cursor over an immutable byte buffer. Never copies.
 *
 * <p>The Java port of the Kanonak wire kernel reader ({@code kanonak.org/wire-form},
 * {@code wireFormatVersion "1"}). {@link #bytes(int)} and {@link #rest()} return
 * read-only {@link ByteBuffer} VIEWS over the backing array — zero-copy; copying
 * is always an explicit caller decision. {@code u32be} widens to {@code long}
 * because Java has no unsigned 32-bit type. UTF-8 decoding is STRICT
 * ({@link CodingErrorAction#REPORT}) — invalid sequences, overlong encodings,
 * and surrogate-range encodings are {@code InvalidUtf8}, never U+FFFD.
 */
public final class WireReader {
    private static final char[] HEX_LOWER = "0123456789abcdef".toCharArray();

    private final byte[] buf;
    private int pos = 0;

    public WireReader(byte[] buf) {
        this.buf = buf;
    }

    private void need(int n, String context) {
        int remaining = buf.length - pos;
        if (remaining < n) {
            throw WireError.truncated(n, remaining, pos, context);
        }
    }

    /** One byte, 0..255. */
    public int u8() {
        need(1, "u8");
        return buf[pos++] & 0xFF;
    }

    /** Two bytes, big-endian, 0..65535. */
    public int u16be() {
        need(2, "u16be");
        int v = ((buf[pos] & 0xFF) << 8) | (buf[pos + 1] & 0xFF);
        pos += 2;
        return v;
    }

    /** Four bytes, big-endian, 0..2^32-1 — widened to {@code long} (Java has no u32). */
    public long u32be() {
        need(4, "u32be");
        long v = (((long) (buf[pos] & 0xFF) << 24)
            | ((buf[pos + 1] & 0xFF) << 16)
            | ((buf[pos + 2] & 0xFF) << 8)
            | (buf[pos + 3] & 0xFF)) & 0xFFFFFFFFL;
        pos += 4;
        return v;
    }

    /** Exactly n bytes as a zero-copy read-only {@link ByteBuffer} view. */
    public ByteBuffer bytes(int n) {
        need(n, "bytes(" + n + ")");
        ByteBuffer v = ByteBuffer.wrap(buf, pos, n).slice().asReadOnlyBuffer();
        pos += n;
        return v;
    }

    /** 16 bytes as a lowercase hyphenated UUID string. Any 16 bytes are legal. */
    public String uuid() {
        need(16, "uuid");
        StringBuilder s = new StringBuilder(36);
        for (int i = 0; i < 16; i++) {
            if (i == 4 || i == 6 || i == 8 || i == 10) {
                s.append('-');
            }
            int b = buf[pos + i] & 0xFF;
            s.append(HEX_LOWER[b >>> 4]).append(HEX_LOWER[b & 0x0F]);
        }
        pos += 16;
        return s.toString();
    }

    /**
     * n bytes decoded as STRICT UTF-8. Bounds are checked before validity
     * ({@code Truncated} beats {@code InvalidUtf8}); the position does NOT
     * advance on a decode failure.
     */
    public String utf8(int n) {
        int start = pos;
        need(n, "bytes(" + n + ")");
        CharsetDecoder decoder = StandardCharsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT);
        try {
            String s = decoder.decode(ByteBuffer.wrap(buf, start, n)).toString();
            pos = start + n;
            return s;
        } catch (CharacterCodingException e) {
            // the read did not take effect — pos stays at start
            throw WireError.invalidUtf8(start, "utf8(" + n + ")");
        }
    }

    /** u16be length L, then exactly L bytes (zero-copy view). L beyond remaining is LengthOverrun. */
    public ByteBuffer lenPrefixedBytes16() {
        int start = pos;
        need(2, "lenPrefixedBytes16");
        int declared = ((buf[pos] & 0xFF) << 8) | (buf[pos + 1] & 0xFF);
        int remainingAfterLength = buf.length - pos - 2;
        if (declared > remainingAfterLength) {
            throw WireError.lengthOverrun(declared, remainingAfterLength, start, "lenPrefixedBytes16");
        }
        pos += 2;
        ByteBuffer v = ByteBuffer.wrap(buf, pos, declared).slice().asReadOnlyBuffer();
        pos += declared;
        return v;
    }

    /** All remaining bytes (possibly empty) as a zero-copy view. Never errors. */
    public ByteBuffer rest() {
        ByteBuffer v = ByteBuffer.wrap(buf, pos, buf.length - pos).slice().asReadOnlyBuffer();
        pos = buf.length;
        return v;
    }

    /** Count of unread bytes. */
    public int remaining() {
        return buf.length - pos;
    }

    /** Errors {@code TrailingBytes} if any bytes remain. */
    public void expectEnd() {
        int count = buf.length - pos;
        if (count > 0) {
            throw WireError.trailingBytes(count, pos);
        }
    }
}
